### ISSUE TITLE
lost request off line

### ISSUE THREAD

**Issue by bill-baroud** [westnordost/StreetComplete#1415]
Sometimes I lost the request when I'm off line. I have to found a new conexion and the points of request reappear.

**How to Reproduce**
- charging requests for an area,
- walk through this area
- and after a few minutes, the points dissappear 

**Versions affected**
Android version 9
StreetComplete version 12.1

**Comment by westnordost** [comment]
Sounds like a bug.

However, I need more information.
I don't understand how this problem is related to being offline or online, can you elaborate?

You must know, once downloaded, the quests are stored offline for up to several weeks, so if the phone loses connection, it should not be relevant.

**Comment by westnordost** [comment]
@bill-baroud ping

**Comment by westnordost** [comment]
No feedback, closing.

**Comment by bill-baroud** [comment]
I'm sorry,
I've not seen the feedback request.
Here the discussion about this problem sent by e-mail a few days ago : 

I believe I've found the problem (it's the same off or on line).
I lost the points of request when the screen turns off (manuelly or by screen saver).
1) request OK
2) screen off + on, no more points
3) soft off + on, points are back
If I turn off the soft and lunch it again with the icon, the points of request display again correctly (even off line).

![1](https://user-images.githubusercontent.com/51325126/60053769-022bb800-96d9-11e9-8c33-0a5c747df2b6.jpg)
![2](https://user-images.githubusercontent.com/51325126/60053770-022bb800-96d9-11e9-90d7-a0a19aa9688f.jpg)
![3](https://user-images.githubusercontent.com/51325126/60053771-02c44e80-96d9-11e9-9f42-5ec2ff93cebe.jpg)
Perhaps you should improve the working by adding a refresh of display when the screen lights again after pause.

**Comment by matkoniecz** [comment]
> screen off + on, no more points

Are you locking phone here? Are you using some password/pin on the lock screen?

**Comment by matkoniecz** [comment]
Some investigation is described in #1456 that is a duplicate of this bug.

I found way to consistently reproduce it on my phone: 

- enable requiring password on unlocking phone
- run SC
- lock phone
- short press power button and wake phone (getting password screen) - according to logcat relay will fail at this time
- type password/give fingerprint

Unlocking phone directly with fingerprint, without password screen, is not causing this bug to manifest.

-----------------------------

Log messages below are using https://github.com/matkoniecz/Zazolc/commits/log_marker_appearance (branching from max weight quest branch)


(1) Locking phone

App stops, listener is set to null in relay within questController property of MainActivity.

```
 E/QuestController: onStop called for QuestController, listener will be unset on the relay
 E/OsmOAuthDialogFragment: listener set to null in VisibleQuestRelay in VisibleQuestRelay with id 1
```

(2) Pressing power button on phone where password is required.

`if(v != null) ViewUtils.postOnLayout(v, this::updateView);` shedules updateView to fire once view is ready in https://github.com/westnordost/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/tangram/MapFragment.java#L193 .

But updateView in QuestMapFragment triggers before listener is set for VisibleQuestRelay.
As result VisibleQuestRelay returns early and `listener.onQuestsCreated(quests, group)` is not called in https://github.com/westnordost/StreetComplete/blob/0457f7f4d4e795ade3f302684ea2216c661a940e/app/src/main/java/de/westnordost/streetcomplete/data/VisibleQuestRelay.java#L25 .

```
 E/OsmOAuthDialogFragment: updateView in QuestMapFragment, will call onFirstInView - what will trigger retrieving quests
 E/QuestController: retrieving quests
 E/QuestController: retrieving quests - relay will be called
 E/OsmOAuthDialogFragment: Threadsafe relay for VisibleQuestListener - onQuestsCreated called, will relay if listener is not null. Relay with id 1
 E/OsmOAuthDialogFragment: listener was null, not relayed. Relay with id 1
 E/QuestController: relay failed
```

After triggering unlock with fingerprint sensor, listener is set.

```
 E/OsmOAuthDialogFragment: onRestart
 E/OsmOAuthDialogFragment: OnStart
 E/OsmOAuthDialogFragment: calling OnStart for questController, listener will be set on the relay
 E/QuestController: onStart called for QuestController, listener will be set on the relay
 E/OsmOAuthDialogFragment: listener set to not null in VisibleQuestRelay in VisibleQuestRelay with id 1
```

But unfortunately it is too late - retrieving quests is not happening again, as result no quests are displayed. Workaround is to force quest fetching again.

**Comment by matkoniecz** [comment]
So how to solve this? One horrific hack that works is along lines

```
				while(!relay.onQuestsCreated(osmQuests, QuestGroup.OSM)) {
					try {
						Thread.sleep(1000);
					} catch (InterruptedException e) {
						Thread.currentThread().interrupt();
					}
				}
```

(I mention it only because it is in a debug code and sort-of works - I used it primarily to confirm that my diagnosis what goes wrong is correct).

Proper fix would be to make `ViewUtils.postOnLayout` sleep as long as it expected (if that is broken) or introduce a new function that will let sleep code longer and use it here.

Or maybe add listener that listenes for setting non-null listener in VisibleQuestRelay?

Or maybe there is no need to actually set listener to null as `onStop` is called for `QuestController`?

I admit that I am no idea how to write it properly. For example - listeners need to be unset in `onStop`, not on `onDestroy` because `onDestroy` may be not called. But what are situation where `onDestroy` of `QuestController` would not be called and applications survises? Is it even possible?

**Comment by westnordost** [comment]
Thank you for the thorough analysis! 

I need to think about this as well. The communication here is really quite complicated. One of the most unclear places in the whole code base, I'd say. So, respect for digging your way through there.

So to get on the same page:
The `MainActivity` sets itself (indirectly) as the listener to be notified of visible quests in `onStart` but the `QuestMapFragment` is starting to be initialized already in `onCreate`. When the `QuestMapFragment` is done with this, it is implicitly assumed that `onStart` of `MainActivity` has been called, too.

So, only if the `QuestMapFragment` is actually created anew, we can run into this race condition. This is of course the case on a new app start, though apparently not a problem, because the first map initialization takes always longer.
But, **question**: When reentering the app, is `QuestMapFragment` really created anew? Because it  shouldn't. And `onSceneReady` in `MapFragment` is only called as a consequence of it being created anew.

So, if my assumptions are correct, the easiest and most correct solution would be for the `MapFragment` to call `updateView` once `onStart`, or maybe for the `MainActivity` to call that on `MapFragment` on its `onStart`, no?

**Comment by matkoniecz** [comment]
> When reentering the app, is QuestMapFragment really created anew? Because it shouldn't.


`QuestsMapFragment` seems to not be created anew, but gets `onStart` and `onResume` calls.


> And `onSceneReady` in `MapFragment` is only called as a consequence of it being created anew.

I set breakpoint there and onSceneReady was triggered only once, at app start.

But https://github.com/matkoniecz/Zazolc/commit/c30bf287ff5efbd0a5d5075f54092d5119981838 confirms that addOnGlobalLayoutListener is triggered many times.

```
07-26 16:17:05.557 31792-31792/de.westnordost.streetcomplete.debug E/OsmOAuthDialogFragment: updateView requested from onSceneReady is source of onGlobalLayout call
07-26 16:17:05.561 31792-31792/de.westnordost.streetcomplete.debug E/OsmOAuthDialogFragment: updateView in QuestMapFragment called
07-26 16:17:05.592 31792-31792/de.westnordost.streetcomplete.debug E/OsmOAuthDialogFragment: updateView requested from onSceneReady is source of onGlobalLayout call
07-26 16:17:05.600 31792-31792/de.westnordost.streetcomplete.debug E/OsmOAuthDialogFragment: updateView in QuestMapFragment called
```

There is `if(vto.isAlive()) vto.removeOnGlobalLayoutListener(this);` that looks like something that should ensure that it is triggered once, but apparently it is not working this way.

**Comment by matkoniecz** [comment]
> call updateView

Note that it requires also modifications to updateView, to prevent

```
		if(lastPos != null  && lastPos.equals(positionNow)) return;
		lastPos = positionNow;
```

add

```
		if(lastDisplayedRect != null && lastDisplayedRect.equals(tilesRect)) return;
```

from returning to early on an additional call. This seems to require propagating success/failure from relay.

**Comment by westnordost** [comment]
OK, so `updateView` is called because of the `postOnLayout`, but we don't know exactly why. Probably because the listener is for some reason not removed after first call.

Then, there are certain checks in `updateView` that prevent the important part of it being called if the view did not change - which is the case when resuming the app.

Let's continue this discussion in this PR: #1500
